### PR TITLE
Transfer cancel link should not appear when the transfer is pending

### PR DESCRIPTION
**Pull request by shoaibmalik786**

### CODE REVIEW

**Review by aquabu — CHANGES_REQUESTED** [review]
Please add rspec test coverage for each conditional branch.

**Comment by shoaibmalik786** [comment]
@aquabu I have written the rspec. Please have a look.

File: app/policies/award_policy.rb [hunk 1]
```diff
@@ -57,7 +57,18 @@ def pay?
   end
 
   def cancel?
-    cancel_by_contributor? || cancel_by_admin?
+    (cancel_by_contributor? || cancel_by_admin?) && status_for_cancel?
+  end
+
+  def status_for_cancel?
```
Comment (aquabu) [inline app/policies/award_policy.rb:63]: The case statement is a convoluted which will make it difficult to maintain. There are also unnecessary code branches that could cause unexpected behavior.

I would recommend refactoring `status_for_cancel?` into a `BlockchainTransactable#cancelable?` method which has object access to `status` and `waiting_in_created?`

That method should have a unit test that covers each branch in the conditional statement.



The

**Review by aquabu — CHANGES_REQUESTED** [review]
Thanks for the additional test. Please refactor the `status_for_cancel?` method.

File: app/policies/award_policy.rb (hunk 1 above)
Comment (shoaibmalik786) [reply app/policies/award_policy.rb]: I have made the changes.

**Review by aquabu — APPROVED** [review]

File: app/models/concerns/blockchain_transactable.rb [hunk 2]
```diff
@@ -67,5 +67,17 @@ def new_blockchain_transaction(params)
         )
       )
     end
+
+    def cancelable?
+      flag = case latest_blockchain_transaction&.status
```
Comment (aquabu) [inline app/models/concerns/blockchain_transactable.rb:72]: @shoaibmalik786 thanks for the refactor. Last minor detail. The case statement can just return directly from `cancelable?`. The `flag = ` and `flag` at the end of the method aren't needed. Thanks!

Comment (shoaibmalik786) [reply app/models/concerns/blockchain_transactable.rb]: I have removed flag.

Comment (aquabu) [reply app/models/concerns/blockchain_transactable.rb]: Awesome. Thanks!

File: app/models/concerns/blockchain_transactable.rb [hunk 3]
```diff
@@ -67,5 +67,16 @@ def new_blockchain_transaction(params)
         )
       )
     end
+
+    def cancelable?
+      case latest_blockchain_transaction&.status
```
Comment (Alexey1100) [inline app/models/concerns/blockchain_transactable.rb:72]: Seems like this doesn't cover all the statuses properly.

Transactable should be cancelable **only** if the latest transaction is created, cancelled or failed, or there are no transactions at all.

**Review by Alexey1100 — CHANGES_REQUESTED** [review]

File: app/models/concerns/blockchain_transactable.rb (hunk 3 above)
Comment (aquabu) [reply app/models/concerns/blockchain_transactable.rb:72]: @Alexey1100 hmm... ok. Let's spec that out in pivotal tracker.

**Pull request merged**